examples/stable_diffusion_xl/gm/models/trainer_factory.py

```diff
@@ -62,70 +65,91 @@ def construct(
 
 
 class LatentDiffusionWithLossGrad(nn.Cell):
-    def __init__(self, network, optimizer, scaler, reducer, overflow_still_update=True, gradient_accumulation_steps=1):
+    def __init__(
+        self,
+        network,
+        optimizer,
+        scaler,
+        reducer,
+        overflow_still_update=True,
+        grad_accum_steps=1,
+        clip_grad=False,
+        clip_norm=1.0,
+    ):
         super(LatentDiffusionWithLossGrad, self).__init__()
         self.grad_fn = ops.value_and_grad(network, grad_position=None, weights=optimizer.parameters)
         self.optimizer = optimizer
         self.scaler = scaler
         self.reducer = reducer
         self.overflow_still_update = overflow_still_update
 
-        assert gradient_accumulation_steps >= 1
-        self.grad_accu_steps = gradient_accumulation_steps
-        if gradient_accumulation_steps > 1:
-            # additionally caches network trainable parameters. overhead caused.
-            # TODO: try to store it in CPU memory instead of GPU/NPU memory.
-            self.accumulated_grads = optimizer.parameters.clone(prefix="grad_accumulated_", init="zeros")
-            self.zeros = optimizer.parameters.clone(prefix="zeros_", init="zeros")
-            self.cur_accu_step = Parameter(Tensor(0, ms.int32), "grad_accumulate_step_", requires_grad=False)
-            self.zero = Tensor(0, ms.int32)
-            for p in self.accumulated_grads:
-                p.requires_grad = False
-            for z in self.zeros:
-                z.requires_grad = False
-        self.map = ops.Map()
-        self.partial = ops.Partial()
+        self.accum_steps = grad_accum_steps
+        self.accum_step = ms.Parameter(ms.Tensor(0, dtype=ms.int32), name="accum_step")
+        self.accumulated_grads = optimizer.parameters.clone(prefix="accum_grad", init="zeros")
+        self.hyper_map = ops.HyperMap()
+        self.clip_grad = clip_grad
+        self.clip_norm = clip_norm
+
+    def do_optim(self, loss, grads):
+        self.accum_step += 1
+        loss = F.depend(
+            loss, self.hyper_map(F.partial(_grad_accum_op, self.accum_steps), self.accumulated_grads, grads)
+        )
+        if self.accum_step % self.accum_steps == 0:
+            if self.clip_grad:
+                grads = ops.clip_by_global_norm(self.accumulated_grads, self.clip_norm)
+                loss = F.depend(loss, self.optimizer(grads))
+            else:
+                loss = F.depend(loss, self.optimizer(self.accumulated_grads))
+            loss = F.depend(loss, self.hyper_map(F.partial(_grad_clear_op), self.accumulated_grads))
+            ops.print("parameters were updated.")
```

Comment:
ops.print may be be necessary and harm the performance

Reply:
delete